nifi/Dockerfile

```diff
@@ -4,74 +4,51 @@
 # hadolint global ignore=DL3038,DL4006
 
-FROM stackable/image/java-base
+FROM stackable/image/java-devel AS builder
 
 ARG PRODUCT
-ARG RELEASE
-
-LABEL name="Apache NiFi" \
-      maintainer="info@stackable.tech" \
-      vendor="Stackable GmbH" \
-      version="${PRODUCT}" \
-      release="${RELEASE}" \
-      summary="The Stackable image for Apache NiFi." \
-      description="This image is deployed by the Stackable Operator for Apache NiFi."
 
 RUN microdnf update && \
     microdnf install \
-    gcc \
-    gzip \
     libxml2-devel \
-    libxslt-devel \
-    tar \
-    unzip \
-    zip && \
-    microdnf install \
-    python-devel \
-    python-pip \
-    python-setuptools && \
+    libxslt-devel && \
     microdnf clean all \
     && rm -rf /var/cache/yum
 
-RUN pip install --no-cache-dir nipyapi==0.19.1
+COPY nifi/licenses /licenses
 
 # The bcrypt tool is needed by NiFi to locally encrypt the admin password that is mounted as a secret in cleartext
 # The source code is available at https://github.com/stackabletech/stackable-utils/tree/main/stackable-bcrypt
 # The binaries are published to Nexus with: mvn clean deploy
 # This jar used to be COPYed into the image from a local binary but it's not possible to do so anymore
 # because /bin is owned by root and curl cannot write there unless it runs as root.
 # This should be fixed by moving the jar to location under /stackable and updating the Nifi operator accordingly.
-RUN curl --fail -L 'https://repo.stackable.tech/repository/m2/tech/stackable/nifi/stackable-bcrypt/1.0-SNAPSHOT/stackable-bcrypt-1.0-20240508.153334-1-jar-with-dependencies.jar' -o /bin/stackable-bcrypt.jar \
-    && chown stackable:stackable /bin/stackable-bcrypt.jar
+RUN curl --fail -L 'https://repo.stackable.tech/repository/m2/tech/stackable/nifi/stackable-bcrypt/1.0-SNAPSHOT/stackable-bcrypt-1.0-20240508.153334-1-jar-with-dependencies.jar' \
+    # This used to be located in /bin/stackable-bcrypt.jar we create a softlink in the main container for backwards compatibility
+    -o /stackable/stackable-bcrypt.jar && \
+    chown stackable:stackable /stackable/stackable-bcrypt.jar
 
 USER stackable
 WORKDIR /stackable
 
-COPY --chown=stackable:stackable nifi/stackable /stackable
-COPY --chown=stackable:stackable nifi/licenses /licenses
-COPY --chown=stackable:stackable nifi/python /stackable/python
-
-
-# zip is different than tar and cannot be just piped, therefore the intermediate save and remove step to unzip
-RUN curl --fail -L https://repo.stackable.tech/repository/packages/nifi/nifi-${PRODUCT}-bin.zip -o /stackable/nifi-${PRODUCT}-bin.zip && \
-    unzip /stackable/nifi-${PRODUCT}-bin.zip && \
-    rm /stackable/nifi-${PRODUCT}-bin.zip && \
-    ln -s /stackable/nifi-${PRODUCT} /stackable/nifi
+RUN curl --fail -L "https://repo.stackable.tech/repository/packages/nifi/nifi-${PRODUCT}-source-release.zip" -o "/stackable/nifi-${PRODUCT}-source-release.zip" && \
+    unzip "nifi-${PRODUCT}-source-release.zip" && \
```

Comment:
there is way too much (and unnecessary) folder juggling here.

in this stage it's fine to just use `/stackable/nifi-{PRODUCT}`
 
also `WORKDIR` is better than `cd`

Reply:
I get that it doesnt look pretty, but there is not much to change i think.
Unzipping the source results in a `nifi-<version>` folder, we want to remove that after the build process but keep the copied binaries that are also `nifi-<version>`. Therefore the suffix `-src`.
Using `WORKDIR` results in another RUN step here as well.

Reply:
Ill try to add some comments to clarify.

Reply:
There is no need to delete anything here.

Just copy from `/stackable/nifi-${PRODUCT}/nifi-assembly/target/nifi-${PRODUCT}-bin/nifi-${PRODUCT}` in the final stage.

The number of steps in the `builder` is (largely) irrelevant

Reply:
I would have agreed and even in the builder image preferred no chaining but only run steps for better debugging. @lfrancke was against it since it creates extra layers and we may hit runner disk limits.

Reply:
Yeah. I did it like @razvan suggested in the past. Optimized builders for readability but we did hit space limits on those intermediate layers on Github runners so I was forced to optimize those again.

NiFi being so large I'm afraid it might happen here as well.

Reply:
IIRC i saw a layer with about 26gb ...